File: 002.Add_Two_Numbers/Solution.py
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        carry = 0
        new_head = None
        curr1, curr2 = l1, l2
        prev, curr = None, None
        while curr1 is not None or curr2 is not None:
            val1, val2 = 0, 0
            if curr1 is not None:
                val1 = curr1.val
                curr1 = curr1.next
            if curr2 is not None:
                val2 = curr2.val
                curr2 = curr2.next
            carry, val = divmod(val1 + val2 + carry, 10)
            curr = ListNode(val)
            if prev is not None:
                prev.next = curr
            if new_head is None:
                new_head = curr
            prev = curr
        if carry > 0:
            curr = ListNode(carry)
            if prev:
                prev.next = curr
        return new_head

File: 002.Add_Two_Numbers/Solution.java
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addTwoNumbersWithCarry(l1, l2, 0);
    }
    
    private ListNode addTwoNumbersWithCarry(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null) {
            if (carry == 0) {
                return null;
            } else {
                return new ListNode(carry);
            }
        }
        int num1 = (l1 == null) ? 0 : l1.val; 
        int num2 = (l2 == null) ? 0 : l2.val;
        ListNode next1 = (l1 == null) ? null : l1.next;
        ListNode next2 = (l2 == null) ? null : l2.next;
        int result = num1 + num2 + carry;
        carry = result / 10;
        result = result % 10;
        
        ListNode l = new ListNode(result);
        l.next = addTwoNumbersWithCarry(next1, next2, carry);
        return l;
    }
}